File: apps/api-go/internal/mailer/mailer.go
package mailer

import (
	"fmt"
	"net/smtp"
	"os"
)

// Service sends via Gmail STARTTLS with an app password (matches the
// NestJS nodemailer `service: 'gmail'` transport). nil = not configured.
type Service struct {
	user     string
	pass     string
	frontend string
}

func New() *Service {
	s := &Service{
		user:     os.Getenv("MAIL_USER"),
		pass:     os.Getenv("MAIL_PASS"),
		frontend: os.Getenv("FRONTEND_URL"),
	}
	if s.user == "" || s.pass == "" {
		return nil
	}
	return s
}

func (m *Service) Enabled() bool { return m != nil }

// FrontendURL returns the configured base URL for building email links.
func (m *Service) FrontendURL() string {
	if m == nil {
		return ""
	}
	return m.frontend
}

func (m *Service) SendVerificationEmail(to, token string) error {
	if m == nil {
		return fmt.Errorf("mailer: MAIL_USER/MAIL_PASS not configured")
	}
	link := m.frontend + "/auth/verify-email?token=" + token
	if m.frontend == "" {
		link = "/auth/verify-email?token=" + token
	}
	body := "From: WebDev Studios <" + m.user + ">\r\n" +
		"To: " + to + "\r\n" +
		"Subject: Verify your email\r\n" +
		"MIME-Version: 1.0\r\n" +
		"Content-Type: text/html; charset=utf-8\r\n\r\n" +
		"<p>Welcome to WebDev Studios!</p>" +
		"<p><a href=\"" + link + "\">Click here to verify your email</a></p>" +
		"<p>This link expires in 24 hours.</p>"
	addr := "smtp.gmail.com:587"
	return smtp.SendMail(addr, smtp.PlainAuth("", m.user, m.pass, "smtp.gmail.com"),
		m.user, []string{to}, []byte(body))
}

// Send is the generic transport used by verification and reset mails alike.
func (m *Service) Send(to, subject, htmlBody string) error {
	if m == nil {
		return fmt.Errorf("mailer: MAIL_USER/MAIL_PASS not configured")
	}
	body := "From: WebDev Studios <" + m.user + ">\r\n" +
		"To: " + to + "\r\n" +
		"Subject: " + subject + "\r\n" +
		"MIME-Version: 1.0\r\n" +
		"Content-Type: text/html; charset=utf-8\r\n\r\n" +
		htmlBody
	return smtp.SendMail("smtp.gmail.com:587", smtp.PlainAuth("", m.user, m.pass, "smtp.gmail.com"),
		m.user, []string{to}, []byte(body))
}

File: apps/api-elysia/src/lib/mailer.ts
import nodemailer from 'nodemailer';

// Gmail STARTTLS with app password, mirrors api-go mailer.go.
// null = not configured; callers degrade (register logs, reset throws 500).

let transport: nodemailer.Transporter | null | undefined;

function getTransport(): nodemailer.Transporter | null {
  if (transport === undefined) {
    transport = null;
    const user = process.env.MAIL_USER;
    const pass = process.env.MAIL_PASS;
    if (user !== undefined && user !== '' && pass !== undefined && pass !== '') {
      transport = nodemailer.createTransport({
        host: 'smtp.gmail.com',
        port: 587,
        secure: false,
        auth: { user, pass },
      });
    }
  }

  return transport;
}

export function frontendUrl(): string {
  return process.env.FRONTEND_URL ?? '';
}

export function mailEnabled(): boolean {
  return getTransport() !== null;
}

export async function send(to: string, subject: string, html: string): Promise<void> {
  const t = getTransport();
  if (t === null) {
    throw new Error('mailer: MAIL_USER/MAIL_PASS not configured');
  }

  await t.sendMail({ from: `WebDev Studios <${process.env.MAIL_USER}>`, to, subject, html });
}

export async function sendVerificationEmail(to: string, token: string): Promise<void> {
  const link = `${frontendUrl()}/auth/verify-email?token=${token}`;
  const body =
    `<p>Welcome to WebDev Studios!</p>` +
    `<p><a href="${link}">Click here to verify your email</a></p>` +
    '<p>This link expires in 24 hours.</p>';
  await send(to, 'Verify your email', body);
}

export async function sendPasswordResetEmail(to: string, token: string): Promise<void> {
  const link = `${frontendUrl()}/auth/reset-password?token=${token}`;
  const body =
    '<p>Reset your password:</p>' +
    `<p><a href="${link}">${link}</a></p>` +
    '<p>Expires in 24 hours.</p>';
  await send(to, 'Password reset', body);
}
